### PR TITLE
Add support for RouteGroups

### PR DESCRIPTION
**Pull request by mikkeloscar**
Adds support for RouteGroups as an alternative to `ingress` or `externalIngress` definitions.

### Migration path

The intended migration path from `Ingress` -> `RouteGroup` or the other way is the following:

1. User adds `RouteGroup` definition and keeps the existing `Ingress` in place. Proposed it to first add the new `RouteGroup` with a different hostname such that it won't overlap with the existing `Ingress` for the initial tests.
2. Once `Ingress` and `RouteGroup` is the same the `Ingress` can be dropped from the `StackSet` spec.
3. stackset-controller will delay the deletion of the `Ingress` if it detects that a `RouteGroup` is defined. This allows the underlying proxy (skipper) to configure the routes for `RouteGroup` before deleting the corresponding routes which were configured for the `Ingress`. - Default setting is to delete the `Ingress` no sooner than 5 minutes after the `RouteGroup` resource has been created. 5 minutes seems like a reasonable time where skipper has plenty of time to be configured AND not too long that it will mess with the routes if the user starts to add additional features via the new `RouteGroup` configuration.

### Traffic weight rounding

Since `RouteGroup`s expect the weight to be in `int` as opposed to `float64` I introduced a rounding which affects the weights when they're calculated to be assigned on the StackSet spec/status thus also changing the behavior slightly for how weights are calculated for `ingress` or `externalIngress`. Basically the weights are still stored in `float64`, but they are rounded to whole numbers still summing up to 100. See the `roundWeights` function in `pkg/core/traffic.go` for details.

This can be the first step for eventually getting rid of floats and just use integers for weights. Related #214 

**WORK IN PROGRESS**

* [x] Make RouteGroup support optional by default
* [x] Gracefully delete ingress if replaced with RouteGroup and vice versa.
* [x] Add Unit tests for RouteGroup
* [x] Add e2e tests for RouteGroup
* [x] Fix traffic distribution (float64 -> int)

### CODE REVIEW

**Review by aryszka — APPROVED** [review]

**Comment by aryszka** [comment]
:+1:

File: pkg/core/traffic.go [hunk 1]
```diff
@@ -46,10 +48,75 @@ func normalizeWeights(backendWeights map[string]float64) {
...
+
+		if fi == fj && ii > ij {
+			return true
+		}
+
+		return fi == fj && ii == ij && weightList[i].Backend < weightList[j].Backend
+	})
+	// check the remaining weight and distribute
+	diff := 100 - sum
+	for idx, backend := range weightList {
```
[the remark this answers is not in the document]
Comment (mikkeloscar) [reply pkg/core/traffic.go]: Nice idea. Changed!

File: pkg/core/traffic.go [hunk 2]
```diff
@@ -89,6 +156,7 @@ func (ssc *StackSetContainer) ManageTraffic(currentTimestamp time.Time) error {
 		} else {
 			normalizeWeights(weights)
 		}
+		roundWeights(weights)
```
[the remark this answers is not in the document]
Comment (mikkeloscar) [reply pkg/core/traffic.go]: :+1:

**Comment by mikkeloscar** [comment]
### Open Question

*  [ ] Should we create default routes for `RouteGroup` if the user doesn't specify anything?

Example:

User just specifies:

```yaml
kind: StackSet
metadata:
  name: stackset-name
spec:
  routegroup:
    hosts: ["example.org"]
    backendPort: 8080
```

And we generate a `RouteGroup`:

```yaml
kind: RouteGroup
metadata:
  name: stackset-name
spec:
  hosts: ["example.org"]
  backends:
  - name: stackset-name-v1
    type: service
    serviceName: stackset-name-v1
    servicePort: 80
  defaultBackends:
  - backendName: stackset-name-v1
    weight: 100
  routes:
  - pathSubtree: /
```

I.e. the routes:

```yaml
  routes:
  - pathSubtree: /
```

Are generated by default so the user has to write less if they just want a simple routing based on hostname (+traffic)

If we go this way, we should ONLY add default `routes` if none are defined.

The **downside** would be that users wanting something like

```yaml
kind: StackSet
metadata:
  name: stackset-name
spec:
  routegroup:
    hosts: ["example.org"]
    backendPort: 8080
    additionalBackends:
    - name: remote
      type: network
      address: https://zalando.de
    routes:
    - pathSubtree: /zalando
      backends:
      - backendName: remote
```

Would need to additionally add the `pathSubtree: /` route.

But the **upside** would be the users can express something like:

```yaml
kind: StackSet
metadata:
  name: stackset-name
spec:
  routegroup:
    hosts: ["example.org"]
    backendPort: 8080
    additionalBackends:
    - name: remote
      type: network
      address: https://zalando.de
    routes:
    - pathSubtree: /api
```

Which would not be possible (result in undesired routing) if we additionally injected  the `pathSubtree: /` route.

**Comment by aryszka** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Comment by mikkeloscar** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Comment by aryszka** [comment]
Some low hanging fruits to improve the coverage:

- the error cases in StackSetController.collectResources()
- the normalizWeight edge cases in traffic.go

**Pull request merged**